File: trunk/icefaces/ace/component/src/org/icefaces/ace/component/textareaentry/TextAreaEntry.java
/*
 * Copyright 2004-2014 ICEsoft Technologies Canada Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the
 * License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

package org.icefaces.ace.component.textareaentry;

import org.icefaces.ace.component.textareaentry.TextAreaEntryBase;
import org.icefaces.ace.event.CharCountEvent;
import org.icefaces.ace.util.Constants;
import org.icefaces.component.Focusable;

import javax.faces.context.FacesContext;
import javax.faces.event.AjaxBehaviorEvent;
import javax.faces.event.FacesEvent;
import java.util.Map;

public class TextAreaEntry extends TextAreaEntryBase implements Focusable {
    public final static String THEME_INPUT_CLASS = "ui-inputfield ui-textareaentry ui-widget ui-state-default ui-corner-all";
    public final static String PLAIN_INPUT_CLASS = "ui-textareaentry";

    public String getFocusedElementId() {
        return getClientId(FacesContext.getCurrentInstance()) + "_input";
    }

    public void queueEvent(FacesEvent event) {
        FacesContext context = FacesContext.getCurrentInstance();
        Map<String,String> params = context.getExternalContext().getRequestParameterMap();
        String eventName = params.get(Constants.PARTIAL_BEHAVIOR_EVENT_PARAM);

        if (eventName != null && eventName.equals("charCount")) {
            if (event instanceof AjaxBehaviorEvent) {
                AjaxBehaviorEvent behaviorEvent = (AjaxBehaviorEvent) event;
                String val = (String) this.getSubmittedValue();
                long currentLength = (val == null || val.isEmpty()) ? 0 : val.length();
                long charsRemaining = this.getMaxlength() - currentLength;
                super.queueEvent(new CharCountEvent(this, behaviorEvent.getBehavior(), currentLength, charsRemaining));
            }
        } else {
            super.queueEvent(event);
        }
    }
}

File: trunk/icefaces/ace/component/resources/icefaces.ace/textareaentry/textareaentry.js
/*
 * Copyright 2004-2014 ICEsoft Technologies Canada Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the
 * License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

ice.ace.TextAreaEntry = function(id, cfg) {
    var jQ = ice.ace.jq;
    var inputId = id + "_input";
    var labelName = id + "_label";
    var maxlength = cfg.maxlength;
    this.id = id;
    this.cfg = cfg;
    this.jqId = ice.ace.escapeClientId(inputId);
    this.jq = jQ(this.jqId);
    var self = this;
	if (cfg.placeholder && !('placeholder' in document.createElement('input'))) { // if 'placeholder' isn't supported, use label inField
		this.cfg.inFieldLabel = this.cfg.placeholder;
	}

    if (cfg.inFieldLabel) {
		if (this.jq.hasClass(cfg.inFieldLabelStyleClass)) {
			this.jq.attr({name: inputId});
			this.jq.val("");
			this.jq.removeClass(cfg.inFieldLabelStyleClass);
		}
        this.jq.focus(
            function() {
                var input = jQ(this);
                if (input.hasClass(cfg.inFieldLabelStyleClass)) {
                    input.attr({name: inputId});
                    input.val("");
                    input.removeClass(cfg.inFieldLabelStyleClass);
                }
            }).blur(
            function() {
                var input = jQ(this);
                if (jQ.trim(input.val()) == "") {
                    input.attr({name: labelName});
                    input.val(cfg.inFieldLabel);
                    input.addClass(cfg.inFieldLabelStyleClass);
                }
            });
    }
    this.jq.blur(function() {
        ice.setFocus();
    });
    this.jq.change(function() {
        ice.setFocus();
    });
    if (maxlength > 0) {
        this.jq.on("keyup change", function (e) {
            var target = e.target;
            if (target.value.length > maxlength) {
                target.value = target.value.substring(0, maxlength);
            }
        });
    }
    if (cfg.behaviors) {
        ice.ace.jq.each(cfg.behaviors, function(name, behavior) {
            if (name == 'charCount') {
                if (document.attachEvent) {
                    //IE 7,8,9 handling -- backspace and delete keypresses do not trigger 'input' events
                    self.jq.on('input', function (e) {
                        if (e.target.value.length <= maxlength) {
                            e.cancelBubble = true;
                            ice.ace.ab(behavior);
                        }
                    });
                    self.jq.on('keyup', function (e) {
                        if (e.target.value.length <= maxlength) {
                            ice.ace.ab(behavior);
                        }
                    });
                } else {
                    self.jq.on('input', function (e) {
                        ice.ace.ab(behavior);
                    });
                }
            } else {
                self.jq.on(name, function () {
                    ice.ace.ab(behavior);
                });
            }
        });
    }

    ice.onElementUpdate(inputId, function() {
        ice.ace.destroy(id);
    });
};

ice.ace.jq(document).on("keydown keypress", function(e){
    if ((e.which || e.keyCode) == 8) {
        var tag = e.target.nodeName;
        if (!('TEXTAREA' == tag || 'INPUT' == tag)) {
            e.preventDefault();
        }
    }
});
